internal/runbits/runtime/progress/progress.go

```diff
@@ -0,0 +1,365 @@
+
+func (a artifactStep) ID() artifactStepID {
+	return artifactStepID(a.artifactID.String() + a.step.String())
+}
+
+type ProgressDigester struct {
+	// The max width to use for the name entries of progress bars
+	maxNameWidth int
+
+	// Progress bars and spinners
+	mainProgress *mpb.Progress
+	buildBar     *bar
+	downloadBar  *bar
+	installBar   *bar
+	solveSpinner *output.Spinner
+	artifactBars map[artifactStepID]*bar
+
+	// Artifact name lookup map
+	artifactNames artifact.Named
+
+	// Track the totals required as the bars for these are only initialized for the first artifact received, at which
+	// time we won't have the totals unless we previously recorded them.
+	buildsExpected    map[artifact.ArtifactID]struct{}
+	downloadsExpected map[artifact.ArtifactID]struct{}
+	installsExpected  map[artifact.ArtifactID]struct{}
+
+	// Debug properties used to reduce the number of log entries generated
+	dbgEventLog []string
+
+	out output.Outputer
+
+	// We use a mutex because whilst this package itself doesn't do any threading; its consumers do.
+	mutex *sync.Mutex
+
+	// The cancel function for the mpb package
+	cancelMpb context.CancelFunc
+
+	// Record whether changes were made
+	changesMade bool
+	// Record whether the runtime install was succesful
+	success bool
+}
+
+func NewProgressDigester(w io.Writer, out output.Outputer) *ProgressDigester {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &ProgressDigester{
+		mainProgress: mpb.NewWithContext(
+			ctx,
+			mpb.WithWidth(progressBarWidth),
+			mpb.WithOutput(w),
+			mpb.WithRefreshRate(refreshRate),
+		),
+
+		artifactNames: map[artifact.ArtifactID]string{},
+		artifactBars:  map[artifactStepID]*bar{},
+
+		cancelMpb:    cancel,
+		maxNameWidth: maxNameWidth(),
+		out:          out,
+
+		mutex: &sync.Mutex{},
+	}
+}
+
+func (p *ProgressDigester) Handle(ev events.Eventer) error {
+	p.dbgEventLog = append(p.dbgEventLog, fmt.Sprintf("%T", ev))
+
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	switch v := ev.(type) {
+
+	case events.Success:
+		p.success = true
+
+	case events.SolveStart:
+		p.solveSpinner = output.StartSpinner(p.out, locale.T("progress_solve"), refreshRate)
+
+	case events.SolveError:
+		if p.solveSpinner == nil {
+			return errs.New("SolveError called before solveBar was initialized")
+		}
+		p.solveSpinner.Stop(locale.T("progress_fail"))
+		p.solveSpinner = nil
+		return nil
+
+	case events.SolveSuccess:
+		if p.solveSpinner == nil {
+			return errs.New("SolveSuccess called before solveBar was initialized")
+		}
+		p.solveSpinner.Stop(locale.T("progress_success"))
+		p.solveSpinner = nil
+		return nil
+
+	case events.BuildSkipped:
+		if p.buildBar != nil {
+			return errs.New("BuildSkipped called, but buildBar was initialized.. this should not happen as they should be mutually exclusive")
+		}
+		return nil
+
+	case events.BuildStarted:
+		p.buildBar = p.addTotalBar(locale.Tl("progress_building", "Building"), v.Artifacts, mpb.BarPriority(StepBuild.priority))
```

Comment:
We should check if buildBar was already started here

Reply:
Do we want to address this? I think there is still a possibility of a `buildBar` check catching out of order events here.